README.md
# *Too many Timestamps*

Simple command line tool to interpret and convert UNIX timestamps

## Usage

### Current time

The actual command is called `timestamps`.
When run with no arguments, it prints info for the current time

![timestamps command showing the current time](timestamps.png)

### Timestamp decoding

To decode a timestamp, simply provide it as argument
```
$ timestamps 1637906683
```

You can optionally decode other formats by providing the corresponding option
```
-S, --seconds         Interpret input as seconds (the default)
-M, --milliseconds    Interpret input as milliseconds
-U, --microseconds    Interpret input as microseconds
-N, --nanoseconds     Interpret input as nanoseconds
```

### Snowflake ID decoding

This tool also supports decoding Discord and Twitter IDs ("Snowflake" format) using these options
```
-D, --discord         Interpret input as a Discord ID
-T, --twitter         Interpret input as a Twitter Snowflake
```
> The non-time info available in Snowflakes is specific to the platform that generated it.


## Building and installing

To build this tool, you should not need anything more than an up to date Rust toolchain and running `cargo build` in the repository.

> I have no way to test this, but it seems that compatibility should be complete for all non-embedded platforms that rust supports.

To install, copy the binary wherever you want on the system, and optionally install the completions for your shell.


## Copyright

Copyright (C) 2021 0x5c

This program is released under the terms of the *Québec Free and Open-Source Licence – Strong Reciprocity (LiLiQ-R+)*, version 1.1.  
See [`LICENCE`](LICENCE) for full license text (Français / English).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).


## [Unreleased]


## [0.1.0] - 2021-11-26
### Added
- Unfinished but functional form of the program


[Unreleased]: https://github.com/0x5c/too-many-timestamps/compare/v0.1.0...HEAD
[0.1.0]: https://github.com/0x5c/too-many-timestamps/releases/tag/v0.1.0


src/types.rs
/*
    Too Many Timestamps
    Copyright (c) 2021 0x5c
    SPDX-License-Identifier: LiLiQ-Rplus-1.1
*/

use std::fmt;

use chrono::{
    DateTime,
    Utc,
};

//#[macro_use]
use crate::parse_snowflake_ts;

pub type DT = DateTime<Utc>;

#[derive(Debug, PartialEq)]
pub enum InputType {
    Seconds,
    Millis,
    Micros,
    Nanos,
    Discord,
    Twitter,
}

impl InputType {
    pub fn from_letter(letter: &str) -> Self {
        match letter {
            "s" => InputType::Seconds,
            "m" => InputType::Millis,
            "u" => InputType::Micros,
            "n" => InputType::Nanos,
            "d" => InputType::Discord,
            "t" => InputType::Twitter,
            &_  => panic!("invalid letter for InputType"),
        }
    }

    pub fn default() -> Self {
        Self::Seconds
    }
}

impl fmt::Display for InputType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let text = match self {
            Self::Seconds => "seconds",
            Self::Millis => "millis",
            Self::Micros => "micros",
            Self::Nanos => "nanos",
            Self::Discord => "Discord ID",
            Self::Twitter => "Twitter ID",
        };
        write!(f, "{}", text)
    }
}

#[derive(Debug)]
pub enum TimeSource {
    Current,
    Input(InputType),
}

impl fmt::Display for TimeSource {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let text = match self {
            Self::Current => String::from("current time"),
            Self::Input(i) => format!("input (as {})", i),
        };
        write!(f, "{}", text)
    }
}

#[derive(Debug)]
pub enum Operation {
    Timestamp(DT),
    Discord(DiscordID),
    Twitter(TwitterSnowflake),
}

#[derive(Debug, PartialEq)]
pub struct DiscordID {
    /// ID creation time
    pub time: DT,
    /// Internal Worker ID
    pub wid: u8,
    /// Internal Process ID
    pub pid: u8,
    /// Sequential
    pub seq: u16,
}

impl DiscordID {
    pub const EPOCH: u64 = 1420070400000;

    pub fn from_int(id: u64) -> anyhow::Result<Self> {
        let time = parse_snowflake_ts!(id, Self::EPOCH);
        let wid = ((id & 0x3E0000) >> 17) as u8;
        let pid = ((id & 0x1F000) >> 12) as u8;
        let seq = (id & 0xFFF) as u16;
        Ok(Self{time, wid, pid, seq})
    }
}

#[derive(Debug, PartialEq)]
pub struct TwitterSnowflake {
    /// Snowflake creation time
    pub time: DT,
    /// Internal machine ID
    pub mid: u16,
    /// Sequential
    pub seq: u16,
}

impl TwitterSnowflake {
    pub const EPOCH: u64 = 1288834974657;

    pub fn from_int(snowflake: u64) -> anyhow::Result<Self> {
        let time = parse_snowflake_ts!(snowflake, Self::EPOCH);
        let mid = ((snowflake & 0x3FF000) >> 12) as u16;
        let seq = (snowflake & 0xFFF) as u16;
        Ok(TwitterSnowflake{time, mid, seq})
    }
}


#[cfg(test)]
mod tests {
    use crate::parse_snowflake_ts;

    use super::{
        InputType,
        DiscordID,
        TwitterSnowflake,
    };

    #[test]
    fn inputtype_from_letter() {
        assert_eq!(InputType::from_letter("s"), InputType::Seconds);
        assert_eq!(InputType::from_letter("m"), InputType::Millis);
        assert_eq!(InputType::from_letter("u"), InputType::Micros);
        assert_eq!(InputType::from_letter("n"), InputType::Nanos);
        assert_eq!(InputType::from_letter("d"), InputType::Discord);
        assert_eq!(InputType::from_letter("t"), InputType::Twitter);
    }

    #[test]
    #[should_panic(expected = "invalid letter for InputType")]
    fn inputtype_from_letter_invalid() {
        let _ = InputType::from_letter("y");  // y tho
    }

    #[test]
    fn discord_from_int() -> anyhow::Result<()> {
        let id: u64 = 564766093051166729;

        let time = parse_snowflake_ts!(id, DiscordID::EPOCH);
        let wid: u8 = 0;
        let pid: u8 = 0;
        let seq: u16 = 9;

        assert_eq!(DiscordID::from_int(id).unwrap(), DiscordID{time, wid, pid, seq});
        
        Ok(())
    }

    #[test]
    fn twitter_from_int() -> anyhow::Result<()> {
        let snowflake: u64 = 1442468859282288643;

        let time = parse_snowflake_ts!(snowflake, TwitterSnowflake::EPOCH);
        let mid: u16 = 343;
        let seq: u16 = 3;

        assert_eq!(TwitterSnowflake::from_int(snowflake).unwrap(), TwitterSnowflake{time, mid, seq});

        Ok(())
    }
}


src/parsing.rs
/*
    Too Many Timestamps
    Copyright (c) 2021 0x5c
    SPDX-License-Identifier: LiLiQ-Rplus-1.1
*/

use std::{
    convert::TryInto,
    num::ParseIntError,
};

use anyhow::bail;
use chrono::{
    Utc,
    DateTime,
    NaiveDateTime,
};

use crate::types::DT;

pub fn parse_input_to_int(input: String) -> Result<i64, ParseIntError> {
    let filterred: String = input.chars().filter(|x| (x != &'.')||(x != &',')).collect();
    filterred.parse()
}

pub enum Split {
    S,
    M,
    U,
    N,
}

/// Split a timestamp into seconds and nanoseconds, at the divisor
/// The multiplier for the nanoseconds part is inversely proprotional to
/// the divisor, so it is derived from it.
pub fn split_timestamp(ts: i64, div: i64) -> anyhow::Result<(i64, u32)> {
    let mul = 1_000_000_000 / div;
    Ok((ts / div, (ts.rem_euclid(div) * mul).try_into()?))
}

pub fn parse_timestamp(ts: i64, split: Split) -> anyhow::Result<DT> {
    let (s, n) = match split {
        Split::S => (ts, 0),
        Split::M => split_timestamp(ts, 1_000)?,
        Split::U => split_timestamp(ts, 1_000_000)?,
        Split::N => split_timestamp(ts, 1_000_000_000)?,
    };
    let ndt = match NaiveDateTime::from_timestamp_opt(s, n) {
        Some(n) => n,
        None => bail!("Timestamp out of range!"),
    };
    Ok(DateTime::<Utc>::from_utc(ndt, Utc))
}

/// Parses the timestamp component out of any ID that follows the Snowflake format
#[macro_export]
macro_rules! parse_snowflake_ts {
    ($snowflake:ident, $epoch:path) => {
        {
            let ts = (($snowflake >> 22) + $epoch) as i64;
            let (sc, ns) = $crate::parsing::split_timestamp(ts, 1_000)?;
            let naive = match chrono::NaiveDateTime::from_timestamp_opt(sc, ns) {
                Some(n) => n,
                None => {
                    println!("{}", "uh oh, the timestamp is out of range, which is likely a bug!\nA bug report will be appreciated :)");
                    panic!("invalid timestamp in snowflake")
                },
            };
            chrono::DateTime::<chrono::Utc>::from_utc(naive, chrono::Utc)
        }
    };
}


#[cfg(test)]
mod tests {
    use chrono::{
        DateTime,
        NaiveDateTime,
        Utc,
    };

    // TODO: parse_input_to_int

    // TODO: split_timestamp

    // TODO: parse_timestamp

    const TWITTER_EPOCH: u64 = 1288834974657;

    #[test]
    fn snowflake_ts_macro() -> anyhow::Result<()> {
        // Actual twitter snowflake found in the wild
        let snowflake: u64 = 1442310554454986761;

        // Directly instanciating a DateTime that should be equivalent
        let ndt = NaiveDateTime::from_timestamp_opt(1632708607, 673000000).unwrap();
        let dt = DateTime::<Utc>::from_utc(ndt, Utc);

        assert_eq!(parse_snowflake_ts!(snowflake, TWITTER_EPOCH), dt);

        Ok(())
    }
}


src/printit.rs
/*
    Too Many Timestamps
    Copyright (c) 2021 0x5c
    SPDX-License-Identifier: LiLiQ-Rplus-1.1
*/

use chrono::SecondsFormat;
use crossterm::style::{
    Stylize,
    style
};

use crate::types::{
    DT,
    DiscordID,
    Operation,
    TimeSource,
    TwitterSnowflake
};


pub fn printit(op: Operation, source: TimeSource) {
    println!("{} {}", "Using".dark_green(), style(source).dark_green());
    match op {
        Operation::Timestamp(time) => print_timestamp(time),
        Operation::Discord(id) => print_discord(id),
        Operation::Twitter(sf) => print_twitter(sf),
    }
}

macro_rules! iso8601 {
    ($time:expr, $format:expr) => {
        println!("{} {}",
            "ISO 8601/RFC 3339:".dark_cyan(),
            $time.to_rfc3339_opts($format, true).blue()
        );
    };
}

macro_rules! rfc2822 {
    ($time:expr) => {
        println!("{} {}", "RFC 2822:".dark_cyan(), $time.to_rfc2822().blue());
    };
}

macro_rules! ts_info {
    ($time:expr, $format:expr) => {
        iso8601!($time, $format);
        rfc2822!($time);
        println!("{}", combo_ts_line($time));
    };
}

fn print_timestamp(time: DT) {
    ts_info!(time, SecondsFormat::AutoSi);
}

fn print_discord(id: DiscordID) {
    ts_info!(id.time, SecondsFormat::AutoSi);
    println!("{} {}", "Discord internal worker ID:".dark_cyan(), style(id.wid).blue());
    println!("{} {}", "Discord internal process ID:".dark_cyan(), style(id.pid).blue());
    println!("{} {}", "Discord internal sequential number:".dark_cyan(), style(id.seq).blue());
}

fn print_twitter(sf: TwitterSnowflake) {
    ts_info!(sf.time, SecondsFormat::AutoSi);
    println!("{} {}", "Twitter internal machine ID:".dark_cyan(), style(sf.mid).blue());
    println!("{} {}", "Twitter internal sequential number:".dark_cyan(), style(sf.seq).blue());
}

fn prepare_combined_ts(time: DT) -> Option<String> {
    let num = format!("{:0>10}", safe_timestamp_nanos(time)?.to_string());
    let (sec, txt) = num.split_at(num.len()-9);
    let (ms, txt2) = txt.split_at(3);
    let (us, ns) = txt2.split_at(3);
    Some(format!("{}{}{}{}", sec.blue(), ms.dark_yellow(), us.red(), ns.dark_magenta()))
}

fn combo_ts_line(time: DT) -> String {
    match prepare_combined_ts(time) {
        Some(ts) => {
            let u = format!("{0}{4}{1}{4}{2}{4}{3}", 
                            "s".blue(),
                            "ms".dark_yellow(),
                            "µs".red(),
                            "ns".dark_magenta(),
                            "/".dark_cyan());
            format!("{} {} {}", "Unix timestamp:".dark_cyan(), ts, u)
        },
        None => format!("{} {}","Unix timestamp:".dark_cyan(), "[time out of bounds]".red()),
    }
}

fn safe_timestamp_millis(dt: DT) -> Option<i64> {
    let as_ns = dt.timestamp().checked_mul(1_000)?;
    as_ns.checked_add(i64::from(dt.timestamp_subsec_millis()))
}

fn safe_timestamp_micros(dt: DT) -> Option<i64> {
    let as_ns = dt.timestamp().checked_mul(1_000_000)?;
    as_ns.checked_add(i64::from(dt.timestamp_subsec_micros()))
}

fn safe_timestamp_nanos(dt: DT) -> Option<i64> {
    let as_ns = dt.timestamp().checked_mul(1_000_000_000)?;
    as_ns.checked_add(i64::from(dt.timestamp_subsec_nanos()))
}


#[cfg(test)]
mod tests {
    use chrono::{
        DateTime,
        NaiveDateTime,
        Utc
    };

    use super::{
        safe_timestamp_millis,
        safe_timestamp_micros,
        safe_timestamp_nanos,
    };

    #[test]
    fn ts_millis() {
        let ndt = NaiveDateTime::from_timestamp_opt(1632802669, 990000000).unwrap();
        let dt = DateTime::<Utc>::from_utc(ndt, Utc);

        assert_eq!(safe_timestamp_millis(dt).unwrap(), 1632802669_990);
    }

    #[test]
    fn ts_micros() {
        let ndt = NaiveDateTime::from_timestamp_opt(1632802669, 990574000).unwrap();
        let dt = DateTime::<Utc>::from_utc(ndt, Utc);

        assert_eq!(safe_timestamp_micros(dt).unwrap(), 1632802669_990_574);
    }

    #[test]
    fn ts_nanos() {
        let ndt = NaiveDateTime::from_timestamp_opt(1632802669, 990574670).unwrap();
        let dt = DateTime::<Utc>::from_utc(ndt, Utc);

        assert_eq!(safe_timestamp_nanos(dt).unwrap(), 1632802669_990_574_670);
    }
}


src/timestamps.rs
/*
    Too Many Timestamps
    Copyright (c) 2021 0x5c
    SPDX-License-Identifier: LiLiQ-Rplus-1.1
*/

use chrono::Utc;
use crossterm::style::{style, Stylize};

use crate::{
    parsing::parse_timestamp,
    types::{
        TimeSource,
        InputType,
        Operation as Op,
        DiscordID,
        TwitterSnowflake,
    },
};

mod printit;
mod args;
mod types;
mod parsing;

fn main() {
    let current_time_at_start = Utc::now();

    let matches = args::create_app();
    
    let (op, source) = if let true = matches.is_present("timestamp") {
        let it = args::find_input_type(&matches);
        let input = match parsing::parse_input_to_int(matches.value_of("timestamp").unwrap().into()) {
            Ok(t) => t,
            Err(_) => {
                eprintln!("{}", "Error: invalid input (digits only, '.,' allowed but ignored)".red());
                std::process::exit(1);
            },
        };
        let op = match dispatch_parsing(input, &it) {
            Ok(o) => o,
            Err(e) => {
                eprintln!("{} {}", "Error:".red(), style(e).red());
                std::process::exit(1);
            }
        };
        (op, TimeSource::Input(it))
    } else {
        (Op::Timestamp(current_time_at_start), TimeSource::Current)
    };

    printit::printit(op, source);
}

fn dispatch_parsing(input: i64, input_type: &InputType) -> anyhow::Result<Op> {
    Ok(match input_type {
        InputType::Seconds => Op::Timestamp(parse_timestamp(input, parsing::Split::S)?),
        InputType::Millis  => Op::Timestamp(parse_timestamp(input, parsing::Split::M)?),
        InputType::Micros  => Op::Timestamp(parse_timestamp(input, parsing::Split::U)?),
        InputType::Nanos   => Op::Timestamp(parse_timestamp(input, parsing::Split::N)?),
        InputType::Discord => Op::Discord(DiscordID::from_int(input as u64)?),
        InputType::Twitter => Op::Twitter(TwitterSnowflake::from_int(input as u64)?),
    })
}


src/args.rs
/*
    Too Many Timestamps
    Copyright (c) 2021 0x5c
    SPDX-License-Identifier: LiLiQ-Rplus-1.1
*/

use clap::{
    App,
    AppSettings,
    Arg,
    ArgMatches,
};

use crate::types::InputType;

pub fn create_app() -> ArgMatches {
    App::new(clap::crate_name!())
            .version(clap::crate_version!())
            .about(clap::crate_description!())
            .after_help("Copyright 2021 0x5c <dev@0x5c.io>\nReleased under the LiLiQ-Rplus-1.1 licence.")
            .override_usage("timestamps (-h | -V)\n    timestamps [[INPUT TYPE] TIMESTAMP]")
            //.setting(AppSettings::ColoredHelp)
            .setting(AppSettings::AllArgsOverrideSelf)
            .setting(AppSettings::DeriveDisplayOrder)
            .arg(Arg::new("timestamp")
                .value_name("TIMESTAMP")
                .about("The timestamp to decode. Defaults to current time.")
                .takes_value(true))
            .help_heading("INPUT TYPE")
            .arg(clap::Arg::new("s")
                .short('S')
                .long("seconds")
                .about("Interpret input as seconds (the default)")
                .group("type"))
            .arg(clap::Arg::new("m")
                .short('M')
                .long("milliseconds")
                .about("Interpret input as milliseconds")
                .group("type"))
            .arg(clap::Arg::new("u")
                .short('U')
                .long("microseconds")
                .about("Interpret input as microseconds")
                .group("type"))
            .arg(clap::Arg::new("n")
                .short('N')
                .long("nanoseconds")
                .about("Interpret input as nanoseconds")
                .group("type"))
            .arg(clap::Arg::new("d")
                .short('D')
                .long("discord")
                .about("Interpret input as a Discord ID")
                .group("type"))
            .arg(clap::Arg::new("t")
                .short('T')
                .long("twitter")
                .about("Interpret input as a Twitter Snowflake")
                .group("type"))
            .group(clap::ArgGroup::new("type")
                .requires("timestamp"))
            .get_matches()
} 

pub fn find_input_type(matches: &ArgMatches) -> InputType {
    match ["s", "m", "u", "n", "d", "t"].iter().find(|x| (matches).is_present(x)) {
        Some(t) => InputType::from_letter(*t),
        None    => InputType::default(),
    }
}
